### PR TITLE
[analysis] Implement Flat lattice

### PR DESCRIPTION
**Pull request by tlively**
Given a type `T`, `Flat<T>` is the lattice where none of the values of `T` are
comparable except with themselves, but they are all greater than a common bottom
element not in `T` and less than a common top element also not in `T`.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6033** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6033" target="_blank">Graphite</a> 
    * **PR #6034** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6034" target="_blank">Graphite</a> 
      * **PR #6035** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6035" target="_blank">Graphite</a> 
        * **PR #6036** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6036" target="_blank">Graphite</a> 
          * **PR #6037** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6037" target="_blank">Graphite</a> 
            * **PR #6038** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6038" target="_blank">Graphite</a> 
              * **PR #6039** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6039" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/6039?utm_source=stack-comment).

**Review by kripken — APPROVED** [review]

File: src/analysis/lattices/flat.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  };
+
+  Element getBottom() const noexcept { return Element{Bot{}}; }
+  Element getTop() const noexcept { return Element{Top{}}; }
+  Element get(T&& val) const noexcept { return Element{std::move(val)}; }
+
+  LatticeComparison compare(const Element& a, const Element& b) const noexcept {
+    if (a.index() < b.index()) {
+      return LESS;
+    } else if (a.index() > b.index()) {
```
Comment (ashleynh) [inline src/analysis/lattices/flat.h:63]: What interface is index() from?

**Review by ashleynh — APPROVED** [review]

File: src/analysis/lattices/flat.h (hunk 1 above)
Comment (tlively) [reply src/analysis/lattices/flat.h:63]: It's built into `std::variant`: https://en.cppreference.com/w/cpp/utility/variant/index

**Comment by tlively** [comment]
### Merge activity

* **Oct 25, 1:00 PM**: @tlively started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/6039).

**Pull request merged**